### PR TITLE
For #13336: Open bookmarks in current tab

### PR DESCRIPTION
**Pull request by czlucius**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Hi there
These are my fixes to #13336, to open bookmarks in the current tab. I am not sure whether this change is approved, but as the issue is still open, I am submitting a draft PR.
May I ask the Fenix UX team for their approval on this change, or is the bookmarks opening in a new tab as intended? I can work on it to add other things if needed.
Thank you!

Here are the videos of the proposed change:

https://user-images.githubusercontent.com/58442255/149168581-54cbf4ab-b79b-4532-9fbb-9881cc96c4cc.mp4

Bookmarklets also work:

https://user-images.githubusercontent.com/58442255/149168733-c9d7f058-a252-4337-8000-aa88877e1d5d.mp4

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt [hunk 1]
```diff
@@ -77,7 +77,7 @@ class DefaultBookmarkController(
         val flags = EngineSession.LoadUrlFlags.select(EngineSession.LoadUrlFlags.ALLOW_JAVASCRIPT_URL)
         openInNewTabAndShow(
             item.url!!,
-            true,
+            false, // See https://github.com/mozilla-mobile/fenix/issues/13336
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt:80]: No need to add a reference to the issue here. The commit message contains a reference to it already.

```suggestion
            false,
```

**Review by jonalmeida — COMMENTED** [review]
The change looks good. You need to fix the tests and comment below. 🙂

**Comment by czlucius** [comment]
@jonalmeida 
I've removed the comment already.
The test  that fails is 
```
TEST: handleBookmarkTapped should open the bookmark
```

According to the test at `BookmarkControllerTest`, it checks if a bookmark is opened in a new tab:
https://github.com/mozilla-mobile/fenix/blob/a49f80d5e8da09968c751b5c3937a5f3dcec71fa/app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkControllerTest.kt#L112-L132

However, this issue is for the bookmark to open in the current tab, so the test is failing

**Comment by jonalmeida** [comment]
That's correct! The verify was asserting that we always opened in a new tab, and this the behaviour you've changed, so we just need to fix the test to verify the new behaviour. 🙂

**Comment by czlucius** [comment]
Ok, I've corrected the test.

**Comment by czlucius** [comment]
About the new behaviour of opening in current tab, since Fenix doesn't associate the home screen with a physical tab, I found out that this will cause a problem: If the user opens a new tab(clicks New Tab), then subsequently opens the bookmark, the bookmark would open in the previous tab, which may not be the intended action the user wanted to take. How should we fix this side effect?
Thanks.

**Comment by czlucius** [comment]
I've modified the test, but it's still failing

**Comment by czlucius** [comment]
@jonalmeida 
May I ask why is the test failing even when both values (the one in the test, and the one in the BookmarkController class) are `false`? I've tried this out on my machine, this does not happen when both are `true`. Is there something else that I am missing here?
Thanks.

**Comment by czlucius** [comment]
I've fixed the tests already.



> About the new behaviour of opening in the current tab, since Fenix doesn't associate the home screen with a physical tab, I found out that this will cause a problem: If the user opens a new tab(clicks New Tab), then subsequently opens the bookmark, the bookmark would open in the previous tab, which may not be the intended action the user wanted to take. How should we fix this side effect? (is there a method to find out whether the user is on the home screen page of a new tab? - there could be other instances where the home screen is launched without the intention to open a new tab, e.g. the home button) Thanks.

What do we do about this?


> handleBookmarkTapped should load the bookmark in a new tab

Should I also change the test name to reflect the new behavior?

Thank you.

**Comment by maverick74** [comment]
@jonalmeida @czlucius isn't there a way to, for example, know if the home button is visible?
If we could know, it could be used to either load the bookmark on a new tab or in the present tab...

**Comment by czlucius** [comment]
@jonalmeida what do I do to continue from here?

**Comment by jonalmeida** [comment]
Sorry, I lost track of the message thread!

> I've fixed the tests already.

🎉 

> > About the new behaviour of opening in the current tab, since Fenix doesn't associate the home screen with a physical tab, I found out that this will cause a problem: If the user opens a new tab(clicks New Tab), then subsequently opens the bookmark, the bookmark would open in the previous tab, which may not be the intended action the user wanted to take. How should we fix this side effect? (is there a method to find out whether the user is on the home screen page of a new tab? - there could be other instances where the home screen is launched without the intention to open a new tab, e.g. the home button) Thanks.
> 
> What do we do about this?

Interesting. There is no concept of a "new tab" in Fenix, it's just the home screen. When the bookmark is opened, does it switch to that tab that it loads in? If so, I think that would be acceptable behaviour for now given how Fenix currently works. We can also discuss this further with UX in a separate issue as well.

> > handleBookmarkTapped should load the bookmark in a new tab
> 
> Should I also change the test name to reflect the new behavior?

Yes please! The name of the test should reflect the behaviour of it. If we're changing the behaviour, someone else coming back to this code in the future will be very confused why it says something and does something else. 🙂

**Comment by czlucius** [comment]
Ok 👍

**Comment by czlucius** [comment]
> Interesting. There is no concept of a "new tab" in Fenix, it's just the home screen. When the bookmark is opened, does it switch to that tab that it loads in? If so, I think that would be acceptable behaviour for now given how Fenix currently works. We can also discuss this further with UX in a separate issue as well.

Your comment got me to think about if the user has 0 tabs(starting from a clean state), then this behaviour will attempt to open the bookmark in a nonexistent current tab, which may result in a crash or no-op. I shall go test this behaviour later.

**Comment by czlucius** [comment]
> > Interesting. There is no concept of a "new tab" in Fenix, it's just the home screen. When the bookmark is opened, does it switch to that tab that it loads in? If so, I think that would be acceptable behaviour for now given how Fenix currently works. We can also discuss this further with UX in a separate issue as well.
> 
> Your comment got me to think about if the user has 0 tabs(starting from a clean state), then this behaviour will attempt to open the bookmark in a nonexistent current tab, which may result in a crash or no-op. I shall go test this behaviour later.

Ironically, the code still works well when I open a bookmark when number of tabs is 0. I guess `openToBrowserAndLoad` already has edge case that settled

**Comment by jonalmeida** [comment]
> Your comment got me to think about if the user has 0 tabs(starting from a clean state), then this behaviour will attempt to open the bookmark in a nonexistent current tab, which may result in a crash or no-op. I shall go test this behaviour later.

This is good train of thought to consider and even better that it is already handled by the `openToBrowserAndLoad` method.

**Comment by jonalmeida** [comment]
@czlucius I tried out the patch and it looks good! However there is one more entry point into bookmarks from the home screen's "Recent bookmarks". I wonder if you're interested in also fixing that too? Otherwise, we can file a separate bug for it.

**Comment by czlucius** [comment]
I think the recent bookmarks from the home screen is somewhat related to this issue: https://github.com/mozilla-mobile/fenix/issues/20012

If bookmarks were to open in the current tab, but top sites still open in a new tab, the behaviour will be quite confusing.

Furthermore, the home screen is also shown when the user explicitly clicks "New tab", and opening the recent bookmark in the current tab would not be the intended behaviour in this case.
Perhaps another issue can be filed about this, and I can look into it?

Thanks.

**Comment by czlucius** [comment]
If the fix is good as it is now, you can merge it.

The minor flaw that I described earlier is illustrated in this recording:

https://user-images.githubusercontent.com/58442255/151705088-f332ad9e-c237-4b40-b2df-e9ca5cab4200.mp4

**Comment by czlucius** [comment]
@jonalmeida 
If the fix is good as it is now, you can merge it(and maybe open a new issue for the home screen one - maybe for top sites, bookmarks and also collections?)

The minor flaw that I described earlier is illustrated in this recording:

https://user-images.githubusercontent.com/58442255/151705088-f332ad9e-c237-4b40-b2df-e9ca5cab4200.mp4

Thanks.

**Comment by Cheap-Skate** [comment]
> @jonalmeida If the fix is good as it is now, you can merge it(and maybe open a new issue for the home screen one - maybe for **top sites, bookmarks and also collections?**)

Please open this issue & Thanks for your efforts so far 👍!!

**Review by jonalmeida — APPROVED** [review]

**Comment by czlucius** [comment]
https://github.com/mozilla-mobile/fenix/issues/2871 is related as well, and since this is fixed, that issue may be fixed too

**Comment by czlucius** [comment]
@jonalmeida 
The issue that I stated above is being reported by some users:
https://github.com/mozilla-mobile/fenix/issues/23595

Is there a method that returns whether the home screen is opened? We could open in new or current tab based on this value.

Thanks!

**Comment by Cheap-Skate** [comment]
#23659 another one reporting the new behavior as a bug

**Comment by czlucius** [comment]
@jonalmeida this seems to be causing problems, as seen in https://github.com/mozilla-mobile/fenix/issues/23595, https://github.com/mozilla-mobile/fenix/issues/23659, and https://github.com/mozilla-mobile/fenix/issues/23746. I think this should be looked into(before this gets into release).

**Comment by maverick74** [comment]
@czlucius i think you missed bug #20012

**Comment by czlucius** [comment]
No, I was talking about the problems caused by changing the behavior to open bookmarks in the current tab.

**Comment by maverick74** [comment]
Oh! My mistake then! Sorry.
But, i think 20012 should be addressed (after the ones you mentioned) as well to avoid more confusion with this change
